### ISSUE TITLE
Performance Issue in 1.2.0

### ISSUE THREAD

**Issue by abri-damien** [Mimetis/Dotmim.Sync#1252]
I have upgraded the NuGet packages from 1.0.2 > 1.2.0 and have noticed a significant degradation in performance. Our integration tests have gone from completing in under 60 seconds to around 5 - 6 minutes.

We are using SQL Server 2022 LocalDb and the SqlSyncChangeTrackingProvider. 

Just wondering if this is a known issue? I can't really justify that big of a decrease in performance so at this stage will have to stay on 1.0.2.

**Comment by VagueGit** [comment]
We noticed a big slowdown on first sync from DMS 1.0.1 to 1.2.0 (SQLite locally to SQL Server via http). The issue went away after the first sync, but the delay in the first sync was so long that we thought many of our customers would have closed the app and contacted our helpdesk. So we took the coward's way out and stayed on 1.0.1

**Comment by Mimetis** [comment]
oh wow, by any chance do you have something you can share to reproduce the issue ?
BTW, can u try the last 1.30 beta version and see if it's still slow ?

**Comment by abri-damien** [comment]
@Mimetis I can try to put something together over the weekend. It still occurs in the latest beta. I have also tested with 1.1.0 and it goes back to normal with that version.

**Comment by Mimetis** [comment]
damn, this commit is pretty simple, and I dont see what could happens here
What is the server provider you are using ?
What are the clients providers you are using ?

**Comment by abri-damien** [comment]
We are using SQL Server 2022 LocalDb and the SqlSyncChangeTrackingProvider. I have tracked it down to this sql command being executed 

![image](https://github.com/user-attachments/assets/fc47cab1-7edd-42fa-ba06-6509f4bd80c9)

using the following approach:

```csharp
var sw = Stopwatch.StartNew();
rowsCleanedCount = await command.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
sw.Stop();
TotalMsRunningCommand += sw.ElapsedMilliseconds;
```

and adding 
`private static long TotalMsRunningCommand { get; set; } = 0;` 
to the top of the file.

This command seems to be what's taking the most time,

**Comment by Mimetis** [comment]
Damn, I don't understand why it's taking time in your case
I've just made some tests and this command, executed on the client, at the end of a sync, is quite instantaneous, every time.
Even after made a lot of changes  on the client.

Do you think you can share a test where you are able to see a big amount of time spent in this command ?

**Comment by abri-damien** [comment]
Apologies for the delay in getting back to you. It could be due to the nature of our testing and the volume of data we are working with. 

Until I get time to prepare something, I can give you a rundown of how the tests are being performed:

1. We are using `WebApplicationFactory<T>` from `Microsoft.AspNetCore.Mvc.Testing` to run the web api in a testable state (using the ASP.NET Core Web Proxy)
2. In the tests that are working with data there's about 3500 records
3. First table has 96 columns and the others each have around 12
4. Theres around 41 tests and out of those 12 are running a sync after data has been inserted/updated/deleted etc

We're doing it this way because our clients have the ability to import large volumes of data at a time, but it doesn't necessarily mean they will.

I'm wondering if there's a way to defer the execution of that command until later or if we can specify an option to skip and then manually run it when we're ready?

**Comment by Mimetis** [comment]
I will remove the auto cleanup routine for sql server change tracking in the next beta version.
For now, disabling the clean up metadata using the `CleanMetadatas` property set to `False` is the right way to go

**Comment by Mimetis** [comment]
If you are using the `SqlSyncChangeTrackingProvider`, the previous versions did not have any algorithm associated with clean metadatas, as it's maintained by **Sql Server** itself (corresponding to `CleanMetadatas=False` in the last beta version)

For all the others providers, the routine is to delete the rows in the `_tracking` tables.

I tried to add a specific call to an internal **Sql Server** stored procedures ("`EXEC sys.sp_flush_CT_internal_table_on_demand;`")  for the `SqlSyncChangeTrackingProvider `provider, to force the clean up but it seems that it's a performance issue now.

In a nutshell, set `CleanMetadatas `to `False` will not change anything compared to the previous version, if you are using `SqlSyncChangeTrackingProvider`.

I will publish a new beta version in the next upcoming days that will remove the call to this routine for the `SqlSyncChangeTrackingProvider` provider, and then we will go back to a previous version without any performance issue.

It will be up to you to have an eye on the change tracking tables and purge them if necessary.

**Comment by Mimetis** [comment]
the last beta has the fix

**Comment by JiriOgurek** [comment]
@Mimetis, same trouble with version 1.2, when I downgrade it's working fine (about 9k records/s comparing to 1k in version 1.2), however when I upgrade to beta 1.3 I am getting this exception... take a look, please.

![Screenshot](https://github.com/user-attachments/assets/a7f55d66-ef1e-45c3-b036-44be44acea95)

**Comment by Mimetis** [comment]
@JiriOgurek, what framework are you using ?

**Comment by JiriOgurek** [comment]
@Mimetis, this is .NET Framework v4.8 project.

**Comment by damien-c-d** [comment]
> @Mimetis, same trouble with version 1.2, when I downgrade it's working fine (about 9k records/s comparing to 1k in version 1.2), however when I upgrade to 1.3.0-beta-1818 I am getting this exception... take a look, please.
> 
> ![Screenshot](https://github.com/user-attachments/assets/a7f55d66-ef1e-45c3-b036-44be44acea95)
> 

You could try adding this to your project file
```xml
<PropertyGroup> <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects> <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType> </PropertyGroup> 
```

Or if that doesn't work you could try adding the binding redirect manually to app.config

```xml
<dependentAssembly>
<assemblyIdentity name="System.ComponentModel.Annotations" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
<bindingRedirect oldVersion="0.0.0.0-4.2.1.0" newVersion="4.2.1.0" />
</dependentAssembly>

```
